lib/storage/sfCacheSessionStorage.class.php

```diff
@@ -29,140 +29,164 @@ class sfCacheSessionStorage extends sfStorage
-      {
-        $data = @unserialize($raw);
-        // We test 'b:0' special case, because such a string would result
-        // in $data being === false, while raw is serialized
-        // see http://stackoverflow.com/questions/1369936/check-to-see-if-a-string-is-serialized
-        if ( $raw === 'b:0;' || $data !== false)
-        {
-          $this->data = $data;
+            if ($signature == sha1("{$id}:{$this->options['session_cookie_secret']}")) {
+                // cookie is valid
+                $this->id = $id;
+            } else {
+                // cookie signature broken
+                $this->id = null;
+            }
+        } else {
+            // cookie format wrong
+            $this->id = null;
         }
-        else
-        {
-          // Probably an old cached value (BC)
-          $this->data = $raw;
+
+        if (empty($this->id)) {
+            $ip = $_SERVER['REMOTE_ADDR'] ?? 'localhost';
+            $ua = $_SERVER['HTTP_USER_AGENT'] ?? 'ua';
+
+            // generate new id based on random # / ip / user agent / secret
+            $this->id = md5(mt_rand(0, 999999) . $ip . $ua . $this->options['session_cookie_secret']);
+
+            if (sfConfig::get('sf_logging_enabled')) {
+                $this->dispatcher->notify(new sfEvent($this, 'application.log', ['New session created']));
+            }
+
+            // only send cookie when id is issued
+            $this->response->setCookie(
+                $this->options['session_name'],
+                "{$this->id}:" . sha1("{$this->id}:{$this->options['session_cookie_secret']}"),
+                $this->options['session_cookie_lifetime'],
+                $this->options['session_cookie_path'],
+                $this->options['session_cookie_domain'],
+                $this->options['session_cookie_secure'],
+                $this->options['session_cookie_httponly']
+            );
+
+            $this->data = [];
+        } else {
+            // load data from cache. Watch out for the default case.
+            $raw = $this->cache->get($this->id, null);
+            if (null === $raw) {
+                $this->data = [];
+            } else {
+                $data = json_decode($raw, true);
+                if (is_array($data)) {
+                    $this->data = $this->upgradeNestedRouteObjects($data);
+                } else {
+                    $data = @unserialize($raw);
+
+                    // We test 'b:0' special case, because such a string would result
+                    // in $data being === false, while raw is serialized
+                    // see http://stackoverflow.com/questions/1369936/check-to-see-if-a-string-is-serialized
+                    if ($raw === 'b:0;' || $data !== false) {
+                        $this->data = $data;
+                    } else {
+                        // Probably an old cached value (BC)
+                        $this->data = $raw;
+                    }
+                }
+            }
```

Comment:
These are the key changes. Because the method is responsible for fetching content from the cache and deserialising it, we need to be mindful that when this first rolls out, the caches will be full of content which has been serialised with `serialize()`.

The introduced changes seek to 
1. first deserialise the content with `json_decode()`, and if that doesn't result in the expected `array`, then
2. fallback to `unserialize()`.

This should mean that any existing sessions can still be decoded, but as you'll see further down in this class, any attempts to (re)cache content will use `json_encode()` instead of `serialize()`.

Additionally, you'll note that if we successfully deserialize with `json_decode()`, that the content is run through `upgradeNestedRouteObject()`. Read on for more detail.